[CUBRIDMAN-182] support with deduplicate clause (#448)
CUBRID/cubrid-manual: en/sql/schema/index_stmt.rst

     
        <index_col_desc2> ::=
                  {
                     (  {column_name | function_name (argument_list)} [ASC | DESC]
                       [{, {column_name | function_name (argument_list)} [ASC | DESC]} ...] )
                  }
                  [WHERE <filter_predicate> ]
                  [WITH <index_with_clause> [{, <index_with_clause>]} ...]
                  [INVISIBLE]
                  [COMMENT 'index_comment_string’]

             <index_with_clause> ::= {ONLINE [PARALLEL parallel_count]} | <index_with_option>
             <index_with_option> ::= {DEDUPLICATE ‘=‘ deduplicate_level}

*   **UNIQUE**: creates an index with unique values.
*   *index_name*: specifies the name of the index to be created. The index name must be unique in the table.

*   *schema_name*: Specifies the schema name. If omitted, the schema name of the current session is used.
*   *table_name*: specifies the name of the table where the index is to be created.
*   *column_name*: specifies the name of the column where the index is to be applied. To create a composite index, specify two or more column names.
*   **ASC** | **DESC**: specifies the sorting order of columns.

> *   *deduplicate_level*: Specifies the deduplicate level (0 to 14). For details, see `DEDUPLICATE`_.

Reviewer: ```suggestion
*   *deduplicate_level*: specifies the deduplicate level (0 to 14). For details, see `DEDUPLICATE`_.
```

Author: thank you.

.. note::

>     *deduplicate_level* is an integer from 0 to 14. 0 means an index with the same configuration as CUBRID 11.2 or earlier without the **DEDUPLICATE** option.

Reviewer: ```suggestion
    *deduplicate_level* is an integer from 0 to 14. 0 means an index with the same configuration as CUBRID 11.2 or earlier which doesn't support the **DEDUPLICATE** option.
```

Author: @mhoh3963 
This sentence will be corrected to accommodate hgryoo's comment above.

If the value of *deduplicate level* is specified as 1 or higher, a hidden index column used internally by the system is added in addition to the index column specified by the user when creating the index. This value is used to reduce redundancy so that it is not biased toward a specific key value. do. The larger the *deduplicate level* value, the more diverse this system value is, and the more relaxed the redundancy is, the shorter the length of the linked list of overflow pages that can be created for a specific key value.

.. note::

>     * In general, the number of key values that can be assigned by the system doubles whenever *deduplicate level*\ increases by one. Therefore, the length of the linked list on the overflow page will be halved.

Reviewer 2: style broken.
```suggestion
    * In general, the number of key values that can be assigned by the system doubles whenever *deduplicate level* increases by one. Therefore, the length of the linked list on the overflow page will be halved.
```

Author: It's corrected. thank you.

    As in the example below, specify the DEDUPLICATE statement directly.
    
    .. code-block:: sql
    
        CREATE TABLE tbl (a int default 0, b int, c int);
        CREATE INDEX i_tbl_b on tbl (b) WITH DEDUPLICATE=3 COMMENT 'for deduplicate level 3';
        CREATE INDEX i_tbl_b_c on tbl (b,c) WITH DEDUPLICATE=7 COMMENT 'for deduplicate level 1';

.. warning::

    * If **deduplicate_key_level** is **\-1**\, it is internally ignored and not applied even if it is specified explicitly. That is, in this case, all indexes are created with *deduplicate level* as **0**.
    * A column name that begins with "**_dedup_**" cannot be created.

.. note::

    * When creating an index, if the composition of the key field is guaranteed to be UNIQUE, the DEDUPLICATE setting set by the user is ignored and *deduplicate level* is created as **0**.
        * When the key field includes all key fields constituting a specific Primary Key or Unique Index
        * However, except when used as an argument of a function

> Duplicated index

Reviewer 2: ```suggestion
Allowing multiple indexes on the same column in the same order
```

"Duplicated index" may cause confusion with the term "deduplicate"

Reviewer 2: Also, this section should be promoted to title level 3 Because the contents are not related to the implicit method or explicit method.

Please refer to the following:
```
Chapter 1 Title
===============

Section 1.1 Title
-----------------

Subsection 1.1.1 Title
~~~~~~~~~~~~~~~~~~~~~~
```

Suggestion:
```
Allowing multiple indexes on the same column in the same order
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Author: OK, thank you.

|                                                                   |                                                                                   |
+-------------------------------------------------------------------+-----------------------------------------------------------------------------------+

.. _deduplicate_overview:

DEDUPLICATE 
-----------

> If you use the **DEDUPLICATE** option, you can improve performance degradation that can occur when modifying index data due to skewed data that is biased toward a specific key value. The value of this option can be adjusted to mitigate the overflow page's linked list for a specific key value from being made too long, thereby improving insert/delete/update and vacuum performance. However, since an index column hidden by the system is added, the number of terminal nodes in the index and the height of the tree may increase, increasing the size of the index and affecting search performance. In particular, if the index data is uniformly distributed with respect to key values, care must be taken because only the size of the index may increase without improving performance.

Reviewer: ```suggestion
If you use the **DEDUPLICATE** option, you can improve performance of modifying skewed index. The value of this option can be adjusted to mitigate the overflow page's linked list for a specific key value from being made too long, thereby improving insert/delete/update and vacuum performance. However, since an index column hidden by the system is added, the number of leaf nodes in the index and the height of the b-tree may increase, increasing the size of the index and affecting search performance. In particular, if the index data is uniformly distributed with respect to key values, you are careful to use this option because only the size of the index may increase without improving performance.
```

Author: Thank you,

> If the value of *deduplicate level* is specified as 1 or higher, a hidden index column used internally by the system is added in addition to the index column specified by the user when creating the index. This value is used to reduce redundancy so that it is not biased toward a specific key value. do. The larger the *deduplicate level* value, the more diverse this system value is, and the more relaxed the redundancy is, the shorter the length of the linked list of overflow pages that can be created for a specific key value.

Reviewer: ```suggestion
If the value of *deduplicate level* is specified as 1 or higher, a hidden index column used internally by the system is added when creating the index. The value of column is used to reduce redundancy so that it is not biased toward a specific key value.  A higher *deduplicate level* value, it will be mitigated redundancy and a smaller overflow page for a specific  key.
```

Author: Thank you.

.. note::

    * In general, whenever *deduplicate level* increases by one, the number of overflow page of specific key will be halved.

>     * The key value given by the system is obtained through the remainder operation for the page number among the OID information of the record. Therefore, the distribution of key values actually given by the system is determined by the number of pages where records are stored.

Reviewer: It seems like a statement that the user doesn't need to know. How about removing this statement ?

Author: OK, I will remove it.
Thank you.

>     * In general, the number of key values that can be assigned by the system doubles whenever *deduplicate level* increases by one. Therefore, the length of the linked list on the overflow page will be halved.

Reviewer: ```suggestion
    * In general, whenever *deduplicate level* increases by one, the number of overflow page of specific key will be halved.
```

Author: Thank you.

>     * The overflow page-linked list is disadvantageous in terms of add, delete, and update performance, but advantageous in terms of retrieval performance. Therefore, it is not desirable to increase the *deduplicate level* to prevent overflow pages from being created. (Adjust the length of the overflow page for one key to be within tens or hundreds of pages)

Reviewer: ```suggestion
    * The overflow page have a disadvantage in add, delete, and update performance, but are beneficial in search performance. Therefore, it is better to adjust the length of the overflow page for one key to be within tens or hundreds of pages, rather than setting the *deduplication level* high so that overflow pages are not created when creating an index with that option.
```

Author: Thank you.

    If **deduplicate_key_level** is greater than or equal to 1, *deduplicate level* is automatically set to **deduplicate_key_level**. 

Explicit method

    This is a method in which the user explicitly specifies the **DEDUPLICATE** option in the SQL statement. Regardless of the **deduplicate_key_level** set value, the *deduplicate level* specified by the user is applied.
    As in the example below, specify the DEDUPLICATE statement directly.
    
    .. code-block:: sql
    
        CREATE TABLE tbl (a int default 0, b int, c int);
        CREATE INDEX i_tbl_b on tbl (b) WITH DEDUPLICATE=3 COMMENT 'for deduplicate level 3';
        CREATE INDEX i_tbl_b_c on tbl (b,c) WITH DEDUPLICATE=7 COMMENT 'for deduplicate level 1';

.. warning::

    * If **deduplicate_key_level** is **\-1**\, it is internally ignored and not applied even if it is specified explicitly. That is, in this case, all indexes are created with *deduplicate level* as **0**.
    * A column name that begins with "**_dedup_**" cannot be created.

.. note::

>     * When creating an index, if the composition of the key field is guaranteed to be UNIQUE, the DEDUPLICATE setting set by the user is ignored and *deduplicate level* is created as **0**.

Reviewer: ```suggestion
    * When creating an index, if the composition of the key field is guaranteed to be UNIQUE, the DEDUPLICATE specified by the user is ignored and *deduplicate level* is created as **0**.
```

>         * When the key field includes all key fields constituting a specific Primary Key or Unique Index

Reviewer: ```suggestion
        * If the columns of index includes all columns of Primary Key or Unique Index.
```

>         * However, except when used as an argument of a function

Reviewer: ```suggestion
        * However, except if columns of Primary key or Unique index used as an argument of a function index.
```

Allowing multiple indexes on the same column in the same order
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        
    Multiple indexes can be defined with all key fields and filter conditions identical except for *deduplicate level*\.

    .. code-block:: sql
    
        CREATE TABLE tbl (a int primary key, b int, c int);
        CREATE INDEX idx1 ON tbl(b, c) WITH DEDUPLICATE=3;
        CREATE INDEX idx2 ON tbl(b, c) WITH DEDUPLICATE=5;
        CREATE UNIQUE INDEX idx_uk ON tbl(b); 
        CREATE INDEX idx3 ON tbl(b, c) WITH DEDUPLICATE=7;

>     In the example above, idx1 and idx2 have the specified *deduplicate level*. However, since idx3 is guaranteed that column b is unique by idx_uk, *deduplicate level* is created as **0**, ignoring the user's designation.

Reviewer: ```suggestion
    In the example above, idx1 and idx2 have the specified *deduplicate level*. However, since idx3 is guaranteed by idx_uk that column b is unique, it is created *duplicate level* as **0** and ignore user's specified option.
```

        
.. note::

>     * If an FK with the same index column differing only in *deduplicate level* already exists, it cannot be duplicated.

Reviewer: ```suggestion
    * If an FK already exists with the same index column that differs only in *deduplication level*, it can't be duplicated.
```

>     * Even if duplicate indexes are allowed, if there is a Primary key or Unique Index with the same configuration, duplicate indexes are not created.

Reviewer: ```suggestion
    * If columns of index includes a primary key or unique index, indexes that differ only in the deduplicate level can't be created. (Even if duplicate indexes are allowed.)
```

    * You cannot change the *deduplicate level* of an index with the ALTER INDEX REBUILD statement. If necessary, delete the index and recreate it.

.. _deduplicate_overflow_page:

OVERFLOW PAGE
~~~~~~~~~~~~~~~

    * What is the overflow page?

>        An index consists of a non-terminal node and a terminal node, and a terminal node is composed of a set of index key information.

Reviewer: ```suggestion
       An index consists of non-leaf nodes and leaf nodes, and a leaf node is composed of a set of index key information.
```

       At this time, one index key information is a pair of a key value and a set of record's OIDs corresponding to the key value.

>        If there are many records with a specific key value, the OIDs of all the records cannot be stored in the terminal node, so they are separated and managed in a separate storage structure, which is called an overflow page.

Reviewer: ```suggestion
       If there are many records with a specific key value, the OIDs of all the records can't be stored in the leaf node, so they are separated and managed in a separate page, which is called an overflow page.
```
